Create Posts “How to choose a third party package” (#22)
nvh95/hung.dev: content/posts/2021-11-21-tăng-tốc.md

### 1. Fit your need
![npm start](/media/npm-start.jpg)

It sounds obvious that you should choose a solution that solves your problem. A package usually tries to solve a set of problems in general. But sometimes, your use case is particular to your situation, and the package might not fit. Please make sure the third party code you are going to install handles sufficient, not too much, not too over engineer your problem. Many times, [DIY](#do-it-yourself) is a good option

### 2. Easy to adopt
![Easy](/media/easy.jpg)

In industry, you don't work alone. So, pick a solution that your team members are able to adapt to quickly. You don't want to choose a solution that you are the only one who can implement and maintain.

### 3. Popularity
![Popularity](/media/popular.jpg)

> It's usually true that a good solution is a popular one. If a solution is good, why isn't it popular? I measure the popularity of a javascript package by the number of github stars and the number of weekly downloads on [https://npmjs.com](https://npmjs.com). A package with a big community proves that it solves the problem for many people. It comes with the advantages of having better support when we run into an issue. The popular package has more maintainers, contributors and users, so it's more mature than the less popular ones. It usually has good documentation and tutorials, which helps you to easily adopt and use it correctly and efficiently.

Reviewer: - If a solution is good, why isn't it popular? => Câu này nên cho vào ngoặc đơn.
- A package with a big community proves that => A package backed by big community shows that
- The popular package has more maintainers => The **more** popular package has more maintainers
- so it's more mature => so it's more complete

> **Bonus:** [https://stateofjs.com/](https://stateofjs.com/) is a great resource for you to know about the trending in javascript every year

Reviewer 2: ```suggestion
**Bonus:** [https://stateofjs.com/](https://stateofjs.com/) is a great resource for you to know about what's trending in javascript each year
```

### 4. In active development
![Team work](/media/teamwork.jpg)

> It's very important that's you are choosing an active project instead of a dead/ unmaintained project. The active project gets better and better over time through community feedback. An unmaintained project does not move forward and does not fix functional bugs and patch security issues. Sometimes, a very popular package could get abandoned and go into the "freeze" state with many open issues and pull requests. It's might used to a great solution in the past, but maybe this is the sign for us we have to move on. An example is [https://github.com/jamiebuilds/react-loadable](https://github.com/jamiebuilds/react-loadable). It's a great solution for a very long time to split code in React. I totally love it. But's stale now with many issues and PRs since 2018 (this post is written at the end of 2021). Now, if I need to split code in React, I rather use https://github.com/gregberge/loadable-components, which is in active development, getting more popular, patch bugs reported by the community and most importantly, solve my problems. My personal advice: choose a package that active in the last 3-6 months, with issues are being resolved and PRs are being merged.

Reviewer: - An unmaintained project does not move forward and does not fix functional bugs and patch security issues. câu này nhiều and quá, đọc hơi mệt, nên dùng dấu phẩy => An unmaintained project does not move forward, does not fix functional bugs, and does not patch security issues.
- 2 link Github nên ghi tên lib rồi nhúng link vào như ở dưới
- It's might used to a great solution in the past => It might be a great solution in the past
- But's stale now => But it's stale now
- that active => that is active/ that has contributions

> It's very important that's you are choosing an active project instead of a dead/ unmaintained project. The active project gets better and better over time through community feedback. An unmaintained project does not move forward,fix functional bugs and patch security issues. Sometimes, a very popular package could get abandoned and go into the "freeze" state with many open issues and pull requests. It's might used to a great solution in the past, but maybe this is the sign for us we have to move on. An example is [react-loadable](https://github.com/jamiebuilds/react-loadable). It's a great solution for a very long time to split code in React. I totally love it. But's stale now with many issues and PRs since 2018 (this post is written at the end of 2021). Now, if I need to split code in React, I rather use [loadable-components](https://github.com/gregberge/loadable-components), which is in active development, getting more popular, patch bugs reported by the community and most importantly, solve my problems. My personal advice: choose a package that active in the last 3-6 months, with issues are being resolved and PRs are being merged.

Reviewer: forward,fix => missing space character

Reviewer 2: ```suggestion
It's very important that you are choosing an active project instead of a dead/unmaintained project. An active project improves over time through community feedback. An unmaintained project does not move forward, fix functional bugs or patch security issues. Sometimes, a very popular package can be abandoned and go into a "frozen" state with many open issues and pull requests. It might have been a great solution in the past, but this is a sign that we have to move on. An example is [react-loadable](https://github.com/jamiebuilds/react-loadable). It was a great solution for a very long time for code-splitting in React. I totally loved it. But it's stale now with many issues and PRs since 2018 (this post is written at the end of 2021). Now, if I need to split code in React, I use [loadable-components](https://github.com/gregberge/loadable-components), which is in active development, becoming more popular, patches bugs reported by the community, and most importantly, solves my problems. My personal advice: choose a package that's active in the last 3-6 months, with issues that are being resolved and PRs that are being merged.
```

### 5. License
![License](/media/license.jpg)

> It's very important that the package you install grant you permission to use it for commercial in production. Packages in the wild are great. But not all of them are free to use. 

Reviewer: - permission => the permission.
- for commercial => for commercial purpose.

> It's very important that the package you install grant you the permission to use it for commercial purpose in production. Packages in the wild are great. But not all of them are free to use. 

Reviewer 2: ```suggestion
It's very important that the package you install grants you the permission to use it for commercial purpose in production. Packages in the wild are great. But not all of them are free to use. 
```

> Let's take an example. If you are working with charts, you might hear of [Highcharts](https://github.com/highcharts/highcharts) - Rock-solid and incredibly flexible charting library made for developers. You download it from [npm](https://www.npmjs.com/), make a chart, it looks great, and you are ready to roll out the feature. But please note that Highcharts is not free for commercial. So if you are writing code for your company, haven't bought a license and use this package for production, you are basically breaking the law. We don't want that, right? So please buy a license before using Highcharts in production. Myself, I use [Recharts](https://github.com/recharts/recharts) instead, a free package powered by [D3](https://github.com/d3/d3) with MIT license and in very active development.

Reviewer: - free for commercial => free for commercial use
- your company, haven't bought => your company, and your company haven't bought

> Software licenses are quite complicated and go beyond the purpose of this post. But in general, package with MIT, Apache 2.0 and BSD are safe to use for commercial. If you choose a package that's are not one of those, please please read the license carefully.

Reviewer: for commercial => for commercial purpose.

### 6. Reasonable size
![Size](/media/size.jpg)

> You don't want to bloat your web application's bundle size by installing a 700kb package. Sometimes you need to, but most of the time, you don't. Please consider the size of the library before making a decision. You need to know about the unpack size, the bundle size, the gzipped size. A larger package means you need more hard disk and bandwidth, which means more money you need to pay, also it takes a longer time for your users to use your app.

Reviewer 2: ```suggestion
You don't want to bloat your web application's bundle size by installing a 700kb package. Sometimes it is necessary, but most of the time it is not. Please consider the size of a library before making a decision. You need to know about the unpacked size, the bundle size, and the gzipped size. A larger package means you need more hard disk and bandwidth, which means more money you need to pay, also it takes a longer time for users to load your app.
```

### 7. DIY
![Do it yourself](/media/diy.jpg)

> As mentioned in [Fit your need](#fit-your-need), many packages try to solve a problem in general (Thus the size of the package is large), you only need a small part of the package. Sometimes, your problem is unique and there are no existing third party packages out there to solve yours. In those cases, it's a great time for you to do it yourself. I find myself in the early days in the industry spending many times finding a third party package to help me build features. But over time, I rarely use external packages for my daily task. It doesn't mean that I always reinvent the wheel. It means that I know what I am doing and I can seek help from the community when I truly need to (for example I will never sanitize user input by myself but use [DOMPurify](https://github.com/cure53/DOMPurify))

Reviewer: - Thus => thus. Đoạn này không viết hoa.
- in the industry => in this industry
- for my daily task => for my daily tasks

> As mentioned in [Fit your need](#fit-your-need), many packages try to solve a problem in general (thus the size of the package is large), you only need a small part of the package. Sometimes, your problem is unique and there are no existing third party packages out there to solve yours. In those cases, it's a great time for you to do it yourself. I find myself in the early days in the industry spending many times finding a third party package to help me build features. But over time, I rarely use external packages for my daily tasks. It doesn't mean that I always reinvent the wheel. It means that I know what I am doing and I can seek help from the community when I truly need to (for example I will never sanitize user input by myself but use [DOMPurify](https://github.com/cure53/DOMPurify))

Reviewer 2: ```suggestion
As mentioned in [Fit your need](#fit-your-need), many packages try to solve a general problem (thus the size of the package is large). You may only need a small part of the package. Sometimes, your problem is unique and there are no existing third party packages out there that solve it. In those cases, it's a great time for you to do it yourself. I found myself in the early days in the industry spending much time finding a third party package to help me build features. But over time, I more rarely used external packages for my daily tasks. It doesn't mean that I always reinvent the wheel. It means that I know what I am doing and I can seek help from the community when I truly need to (for example I will never sanitize user input by myself, but use [DOMPurify](https://github.com/cure53/DOMPurify))
```

> **Bonus:** For some problems, if you cannot find a package helps you, just make one and give it to the world by making an open source software. You will receive more than you can think of.

Reviewer 2: ```suggestion
**Bonus:** For some problems, if you cannot find a package that helps you, make one and give it to the world by making it an open source software.
```

Reviewer 2: Deleted that last sentence because it doesn't make sense, but it could be replaced with something else!

## Conclusion

> Above are my very personal tips, it might be true and it might not be true. If there's something you don't agree with, that's OK. Just ignore it, or better, let me know why you don't agree. If you have additional tips, please let me know in the comment section. I would love to hear from you. I wish you can choose the right packages for you and your team and happy coding.

Reviewer: for you and your team and happy coding. => for you and your tea. Happy coding!

> Above are my very personal tips, it might be true and it might not be true. If there's something you don't agree with, that's OK. Just ignore it, or better, let me know why you don't agree. If you have additional tips, please let me know in the comment section. I would love to hear from you. I wish you can choose the right packages for you and your team. Happy coding!

Reviewer 2: ```suggestion
Above are my very personal tips, they might be true and they might not be true. If there's something you don't agree with, that's OK. Just ignore it, or better, let me know why you don't agree. If you have additional tips, please let me know in the comment section. I would love to hear from you. I hope this helps you choose the right packages for you and your team. Happy coding!
```
